### PR TITLE
Tree: add new auto check mode

### PR DESCRIPTION
**Pull request by Knicksaas**
363033

### CODE REVIEW

File: eclipse-scout-core/src/tree/Tree.ts [hunk 1]
```diff
@@ -230,8 +230,9 @@ export class Tree extends Widget implements TreeModel {
 
   /**
    * Initialize nodes, applies filters and updates flat list
+   * @internal
    */
-  protected _initNodes(nodes: TreeNode[], parentNode?: TreeNode) {
+  _initNodes(nodes: TreeNode[], parentNode?: TreeNode) {
```
Comment (cguglielmo) [inline eclipse-scout-core/src/tree/Tree.ts:235]: Don't remove protected key word only for tests. You can do it in the SpecTree

File: eclipse-scout-core/src/tree/Tree.ts [hunk 2]
```diff
@@ -299,6 +300,17 @@ export class Tree extends Widget implements TreeModel {
     ]);
   }
 
+  /**
+   * @param autoCheckChildren
```
Comment (cguglielmo) [inline eclipse-scout-core/src/tree/Tree.ts:304]: remove empty js docs

File: eclipse-scout-core/src/tree/TreeModel.ts [hunk 3]
```diff
@@ -16,7 +16,9 @@ export interface TreeModel extends WidgetModel {
   toggleBreadcrumbStyleEnabled?: boolean;
   breadcrumbTogglingThreshold?: number;
   /**
-   * Default is false.
+   * Describes the behavior of children and parent nodes, when a node is checked/unchecked
```
Comment (cguglielmo) [inline eclipse-scout-core/src/tree/TreeModel.ts:19]: What is the behavior? Describe the behavior -> you can copy it from AbstractTree.getConfiguredAutoCheckChildren

File: eclipse-scout-core/test/tree/TreeAdapterSpec.ts [hunk 4]
```diff
@@ -236,7 +236,7 @@ describe('TreeAdapter', () => {
 
   describe('checkNodes', () => {
 
-    it('does not send checked event if triggered by server', () => {
+    it('does send checked event of checked node if triggered by server', () => {
```
Comment (cguglielmo) [inline eclipse-scout-core/test/tree/TreeAdapterSpec.ts:239]: remove "does". Just write "sends". The same applies for every test case

File: eclipse-scout-core/src/tree/TreeCheckNodesResult.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2010, 2023 BSI Business Systems Integration AG
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+import {TreeNode} from './TreeNode';
```
Comment (cguglielmo) [inline eclipse-scout-core/src/tree/TreeCheckNodesResult.ts:10]: Always import from index

File: eclipse-scout-core/test/tree/TreeSpec.ts [hunk 6]
```diff
@@ -1278,6 +1277,407 @@ describe('Tree', () => {
       }, node.childNodes);
     });
 
+    describe('with children_and_parent AutoCheckStyle', () => {
```
Comment (cguglielmo) [inline eclipse-scout-core/test/tree/TreeSpec.ts:1280]: change to describe('autoCheckChildren')

File: eclipse-scout-core/test/tree/TreeSpec.ts [hunk 7]
```diff
@@ -1244,12 +1244,12 @@ describe('Tree', () => {
       expect(checkedNodes.length).toBe(1);
     });
 
-    it('checks children if autoCheckChildren is set to true', () => {
+    it('checks children if tree is in autoCheckChildren mode', () => {
```
Comment (cguglielmo) [inline eclipse-scout-core/test/tree/TreeSpec.ts:1247]: Are these two tests still necessary? Or are they replaced anyway by the new tests below?

File: eclipse-scout-core/test/tree/TreeSpec.ts [hunk 8]
```diff
@@ -1278,6 +1277,407 @@ describe('Tree', () => {
...
+        expect(tree.checkedNodes).toEqual([]);
+        expect(parentNode.checked).toBe(false);
+        expect(parentNode.childrenChecked).toBe(false);
+        expect(childNode1.checked).toBe(false);
+        expect(childNode1.childrenChecked).toBe(false);
+        expect(childNode2.checked).toBe(false);
+        expect(childNode2.childrenChecked).toBe(false);
+      });
+
+      it('does preserve the setted value after init', () => {
```
Comment (cguglielmo) [inline eclipse-scout-core/test/tree/TreeSpec.ts:1559]: setted -> set

File: eclipse-scout-core/test/tree/TreeSpec.ts [hunk 9]
```diff
@@ -1278,6 +1277,407 @@ describe('Tree', () => {
...
+        expect(insertedNode.checked).toBe(false);
+      });
+
+      const ensureRenderedCorrectly = (nodes: TreeNode[]) => {
+        Tree.visitNodes(node => {
+          if (!node.$node) {
+            return;
+          }
+
+          let hascheckedCssClass = node.$node
```
Comment (cguglielmo) [inline eclipse-scout-core/test/tree/TreeSpec.ts:1665]: -> hasChecked (uppercase c)

File: org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/ui/basic/tree/AbstractTree.java [hunk 10]
```diff
@@ -2768,7 +2747,8 @@ public void setNodesCheckedFromUI(List<ITreeNode> nodes, boolean checked) {
           setTreeChanging(true);
           nodes = resolveNodes(nodes);
           if (!nodes.isEmpty()) {
-            setNodesChecked(nodes, checked, true);
+            // With isAutoCheckChildNodes() it's possible to check disabled nodes from UI
```
Comment (cguglielmo) [inline org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/ui/basic/tree/AbstractTree.java:2750]: Why is this necessary?

File: eclipse-scout-core/src/form/fields/treebox/TreeBox.ts [hunk 11]
```diff
@@ -79,7 +79,7 @@ export class TreeBox<TValue> extends LookupBox<TValue> implements TreeBoxModel<T
     this._valueSyncing = true;
     let opts = {
       checkOnlyEnabled: false,
-      checkChildren: false
+      checkChildren: this.tree.autoCheckChildren
```
Comment (cguglielmo) [inline eclipse-scout-core/src/form/fields/treebox/TreeBox.ts:82]: Remove checkChildren here and check for this.tree.autoCheckChildren below. No need to pass the default value of an option

**Review by cguglielmo — APPROVED** [review]

File: org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/ui/basic/tree/AbstractTree.java (hunk 10 above)
Comment (Knicksaas) [reply org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/ui/basic/tree/AbstractTree.java:2750]: Until now, it was not possible to check a disabled node from UI. Since the autoCheckChildren-change, disabled nodes can be checked by checking all child nodes of a disabled node. I have added more details to the comment.
